keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -221,18 +242,27 @@ class SqlArtifactRepository(
     val environment = deliveryConfig.environmentNamed(targetEnvironment)
     val envUid = deliveryConfig.getUidFor(environment)
     val artifactId = artifact.uid
-    val versions: List<String> = sqlRetry.withRetry(READ) {
-      jooq
-        .select(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION)
-        .from(ENVIRONMENT_ARTIFACT_VERSIONS)
-        .where(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(envUid))
-        .and(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID.eq(artifactId))
-        .orderBy(ENVIRONMENT_ARTIFACT_VERSIONS.APPROVED_AT.desc())
-        .fetch()
-        .getValues(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION)
-    }
 
-    return versions.firstOrNull()
+    /**
+     * If [targetEnvironment] has been pinned to an artifact version, return
+     * the pinned version. Otherwise return the most recently approved version.
+     */
+    return sqlRetry.withRetry(READ) {
+      jooq.select(ENVIRONMENT_ARTIFACT_PIN.ARTIFACT_VERSION)
+        .from(ENVIRONMENT_ARTIFACT_PIN)
+        .where(
+          ENVIRONMENT_ARTIFACT_PIN.ENVIRONMENT_UID.eq(envUid),
+          ENVIRONMENT_ARTIFACT_PIN.ARTIFACT_UID.eq(artifactId))
+        .fetchOne(ENVIRONMENT_ARTIFACT_PIN.ARTIFACT_VERSION)
+        ?: jooq
```

Comment:
This change to `latestVersionApprovedIn` ensures cluster resources use a pinned artifact version, if present, when calculating desired state. This is applicable for `DockerImageResolver` and the EC2 `ImageResolver`, with the exception of Jenkins Artifacts which are only supported by the EC2 `ImageResolver`.  I haven't thought through pinning for jenkins artifacts yet.